flow/aggregates.py

```diff
@@ -0,0 +1,481 @@
+        aggregates, aggregate_ids = make_aggregates()
+        self._aggregates = aggregates
+        self._aggregate_ids = aggregate_ids
+
+
+class _DefaultAggregateStore:
+    """This class holds the information of the project associated with
+    an operation function aggregated by the default aggregates i.e.
+    :py:class:`aggregator.groupsof(1)`.
+
+    When iterated through it's instance, it yields a tuple of a single job from
+    the Project.
+
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, project):
+        self._project = project
+
+    def __iter__(self):
+        for job in self._project:
+            yield (job,)
+
+    def __getitem__(self, id):
+        "Return a tuple of a single job via job id."
+        try:
+            return (self._project.open_job(id=id),)
+        except KeyError:
+            raise LookupError(f"Did not find job with id {id}.")
+
+    def __contains__(self, job):
+        """Return whether the job is present in the project associated with this
+        instance of :py:class:`_DefaultAggregateStore`.
+        """
+        return job in self._project
+
+    def __len__(self):
+        return len(self._project)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._project == other._project
+
+    def __hash__(self):
+        blob = self._project.id
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _register_aggregates(self, project):
+        """We have to store self._project when this method is invoked
+        This is because we will then iterate over that project in
+        order to return an aggregates of one.
+        """
+        self._project = project
+
+
+class _MakeAggregates:
+    """This class acts as a funnctor which handles the creation of
+    aggregates using all the jobs in a project.
+
+    :param aggregate:
+        aggregator object associated with this class.
+    :type aggregate:
+        :py:class:`aggregator`
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, aggregate, project):
+        self._aggregate = aggregate
+        self._project = project
+
+    def __call__(self):
+        jobs = list(self._project)
+
+        if self._aggregate._select is not None:
+            jobs = filter(self._aggregate._select, jobs)
+        if self._aggregate._sort_by is not None:
+            jobs = sorted(jobs,
+                          key=lambda job: job.sp[self._aggregate._sort_by],
+                          reverse=self._aggregate._reverse_order)
+        aggregated_jobs = self._aggregate._aggregator_function(jobs)
```

Comment:
I may have been confusing here. If we just use `filter` without `sorted` then `jobs` will not be a list when passed to `self._aggregate._aggregator_function`. It will be a `filter` object, so if `sort_by` is `None` then we do need to explicitly cast to a list, if that is an API guarantee to the user for the aggregation function rather than promising to return an iterable of jobs.

Reply:
So, I should probably execute `jobs = list(jobs)` after we filter jobs